wasm-chain-integration/src/lib.rs

```diff
@@ -284,10 +292,18 @@ pub struct InitHost<'a> {
     /// The parameter to the init method.
     pub param:             &'a [u8],
     /// The init context for this invocation.
-    pub init_ctx:          &'a InitContext<&'a [u8]>,
+    pub init_ctx:          &'a Ctx,
 }
 
-pub struct ReceiveHost<'a> {
+pub struct ReceiveHost<'a, Ctx, P, A>
+where
+    A: AsRef<[u8]>,
+    P: SerialPolicies<A>,
+    Ctx: HasReceiveContext<P>, {
+    /// Phantom marker for A.
+    policy_bytes_type:     std::marker::PhantomData<A>,
+    /// Phantom marker for P.
+    policy_type:           std::marker::PhantomData<P>,
```

Comment:
And the same here.

You just require the constraints when you are using the type. I think that generally works better, and we very rarely have constraints on data type definitions.